File: parnamondal/Week - 2/Day - 1/Sort Integers by The Number of 1 Bits.py
# Time Complexity - O(nlogn)
# Space Complexity - O(n)


class Solution:
    def sortByBits(self, arr: List[int]) -> List[int]:
        n=len(arr)
        count=[]
        for i in arr:
            c=bin(i)[2:].count('1')
            count.append([c,i])
        count.sort()
        #print (count)
        n=[]
        for i in count:
            n.append(i[1])
        return n

File: parasmittal/Day 7 & Day 8/Sort Integers by The Number of 1 Bits.cpp
class Solution {
public:
    int NumberOfBits(int val){
        int count=0;
        for(int i=0;i<32;i++){
            if((val>>i)%2)
                count++;        }
        return count;
    }
    vector<int> sortByBits(vector<int>& arr) {
        unordered_map<int,vector<int>> um;
        int n=arr.size();
        if(n<=1)
            return arr;
        for(int i=0;i<n;i++)
        {
            int number=NumberOfBits(arr[i]);
            um[number].push_back(arr[i]);
        }
        vector<int> ans;
        for(int i=0;i<31;i++){
            if(um.find(i)!=um.end()){
                sort(um[i].begin(),um[i].end());
                for(int j=0;j<um[i].size();j++)
                    ans.push_back(um[i][j]);
            }
        }
        return ans;
    }
};